src/pages/LogOutPreviousUserPage.js

```diff
@@ -0,0 +1,51 @@
+import React, {Component} from 'react';
+import lodashGet from 'lodash/get';
+import PropTypes from 'prop-types';
+import {withOnyx} from 'react-native-onyx';
+import ONYXKEYS from '../ONYXKEYS';
+import * as Session from '../libs/actions/Session';
+import FullScreenLoadingIndicator from '../components/FullscreenLoadingIndicator';
+import Navigation from '../libs/Navigation/Navigation';
+
+const propTypes = {
+    /** The parameters needed to authenticate with a short lived token are in the URL */
+    route: PropTypes.shape({
+        /** Each parameter passed via the URL */
+        params: PropTypes.shape({
+            /** The email of the transitioning user */
+            email: PropTypes.string,
+        }),
+    }).isRequired,
+
+    /** The data about the current session which will be set once the user is authenticated and we return to this component as an AuthScreen */
+    session: PropTypes.shape({
+        /** The user's email for the current session */
+        email: PropTypes.string,
+    }).isRequired,
+};
+
+class LogOutPreviousUserPage extends Component {
+    componentDidMount() {
+        const paramsEmail = lodashGet(this.props, 'route.params.email', null);
+        const sessionEmail = lodashGet(this.props.session, 'email', '');
+        if (paramsEmail !== sessionEmail) {
+            Session.signOutAndRedirectToSignIn();
+            return;
+        }
+
+        // Set isNavigationReady so that we can navigate in the AuthScreens
+        Navigation.setIsNavigationReady();
```

Comment:
This is the part that feels really dangerous to me. How does anyone know when to call this and from where? If it's missed, it will lead to bugs, and if people don't understand this properly, it will lead to people putting this all over the code (creating an unnecessary mess).

I know you explored adding this to its own mounted component, but I lost the comment thread on why that didn't work or what problems you ran into that caused it to be abandoned. Can you link to that by any chance?

At the very least, I would ask to have a very explicit comment about why this is here and why it shouldn't be anywhere else.

Reply:
> This is the part that feels really dangerous to me. How does anyone know when to call this and from where? If it's missed, it will lead to bugs, and if people don't understand this properly, it will lead to people putting this all over the code (creating an unnecessary mess).

Yeah fair point. Hopefully people won't have to use this too much if at all.

> I know you explored adding this to its own mounted component, but I lost the comment thread on why that didn't work or what problems you ran into that caused it to be abandoned. Can you link to that by any chance?

I think the problem was that we can't include the `<NavigationReadyDetector />` inside the StackNavigator because according to the error message "A navigator can only contain 'Screen', 'Group' or 'React.Fragment' as its direct children". 
<img width="1704" alt="image" src="https://user-images.githubusercontent.com/26260477/169890589-66994e54-0d19-434d-bcdf-2a881530f633.png">

If we include in in the `AppNavigator` as I did [here](https://github.com/Expensify/App/blob/468c5a053140103bad2723529891f6f398e2bdc5/src/libs/Navigation/AppNavigator/index.js#L17-L20) then it also doesn't work because it is mounted before the navigation is ready.

The `onReady` prop of the NavigationContainer doesn't work either because it only runs when the navigation container mounts for the first time, which is as soon as Expensify.js renders. We discussed that [here](https://github.com/Expensify/App/pull/8855#discussion_r863133914).

The only other thing that I can think of doing is using the parseAndStoreRoute function to update `isNavigationReady`. If there's state then it's ready, otherwise it's not. I'll try that out.
https://github.com/Expensify/App/blob/cd397ae9aa6f7f2e2cb1910da3f1b28ce2147882/src/libs/Navigation/NavigationRoot.js#L37-L55

> At the very least, I would ask to have a very explicit comment about why this is here and why it shouldn't be anywhere else.
I will add that if I can't get my solution above to work.

Reply:
>The only other thing that I can think of doing is using the parseAndStoreRoute function to update isNavigationReady. If there's state then it's ready, otherwise it's not. I'll try that out.

Sounds like a good idea though `PublicScreens` also probably generates these state change events.

Reply:
I think I got that solution to work. I don't see it being a problem if `PublicScreens` also changes the state.